Feature: Soy Views Renderer

sgammon opened the pull request:
This changeset introduces a new Micronaut package to render views via *[Soy](https://github.com/google/closure-templates)* (also known as *Closure Templates*).

For now, the renderer simply uses `SoyTofu` to render views from Soy template sources. There are more efficient ways to render Soy templates in Java, and I'm planning to add those, too, subject to feedback from the Micronaut team.

I've based this proposal on the structure of `views-freemarker`, to make sure I've adopted Micronaut's coding style as closely as I can. There is a corresponding set of unit tests as well, that pass, at least locally.

_NOTE:_ This change requires a patched version of Soy Java, which applies a PR upstream to support querying a `SoyTofu` instance for templates (see: google/closure-templates#185). A patched JAR for Soy can be installed with the script `patch-soy.sh` added at the root of the repo. This is run in Travis, for instance, but can be eliminated when the upstream PR is merged and a release is cut.

Changes made:
- [x] Support for `SoyTofu` - i.e. rendering from template sources
  - [x] `SoyTofuRenderer` - where the rendering happens
  - [x] `SoyTofuViewsRendererConfiguration` - where a `SoyFileSet` is loaded and configured
  - [x] `SoyFileSetProvider` - uses DI to resolve a `SoyFileSet`, which is used to back the `SoyTofu.Renderer`
- [x] Unit tests for `SoyTofuViewsRenderer`
- [x] `SoySauce` - i.e. rendering via templates pre-compiled to Java sources
- [x] Extra support for Soy features
  - [x] Async continuation during render (filed as #15)
  - [x] Injection of CSP nonces - filed separately as #16, because it depends on #14
  - [x] Support for rewrite maps

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/micronaut-projects/micronaut-views?pullRequest=13) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/micronaut-projects/micronaut-views?pullRequest=13) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/micronaut-projects/micronaut-views?pullRequest=13) it.</sub>

sgammon commented:
It took a few tries to get CI working, due to the JAR injection. Obviously it may be best to wait for the upstream PR to be merged which would allow this to be much cleaner.

sgammon commented:
It took a few tries to get CI working, due to the JAR injection. Obviously it may be best to wait for the upstream PR to be merged which would allow this to be much cleaner. Test coverage for the added code stands at 100%.

What do you think, @graemerocher? :)

graemerocher commented on views-soy/src/main/java/io/micronaut/views/soy/AppendableToWritable.java:
```diff
@@ -0,0 +1,44 @@
+package io.micronaut.views.soy;
```
Missing license declaration

sgammon replied:
Good catch will fix

graemerocher commented on views-soy/src/main/java/io/micronaut/views/soy/AppendableToWritable.java:
```diff
@@ -0,0 +1,44 @@
+
+
+/**
+ * Adapts {@link Appendable} to {@link Writable} for use when rendering Soy templates.
```
Missing javadoc author/version tags

sgammon replied:
ty will fix

graemerocher commented on views-soy/src/main/java/io/micronaut/views/soy/SoyFileSetProvider.java:
```diff
@@ -0,0 +1,29 @@
+package io.micronaut.views.soy;
```
Missing license statement

graemerocher commented on views-soy/src/main/java/io/micronaut/views/soy/SoySauceViewsRenderer.java:
```diff
@@ -0,0 +1,134 @@
+package io.micronaut.views.soy;
```
Missing license statement

sgammon commented:
By the way, we have been testing this in production for a few days now. Simple page but it serves templates. Here's what performance looks like under some basic load:

<img width="531" alt="Screen Shot 2019-09-02 at 10 34 18 AM" src="https://user-images.githubusercontent.com/171897/64128792-412b5900-cd6d-11e9-83d1-48aca684fc0a.png">

this is why we're so excited for Soy in micronaut :)

sgammon commented:
just requested a release at google/closure-templates#188, will update here when no override JAR dependencies are needed.

sgammon commented:
@graemerocher the following is done:

- [x] Soy was released upstream, I've removed the JARs and updated Gradle deps
- [x] Added missing copyright headers
- [x] Fixed author/version tags (marked as `@since 1.2.1`, since that's the current snapshot version)

graemerocher commented:
Ok thanks, I think the only 2 changes necessary left are to cleanup the repos here:

https://github.com/micronaut-projects/micronaut-views/pull/13/files#diff-b1c3b945cac0534cb40fc83347c5f93cR1

Subprojects shouldn't have additional repos, I am not sure what these repos are necessary for. If an additional one is necessary please add it to the root project.

And finally, we will need some documentation added to the guide for Soy https://github.com/micronaut-projects/micronaut-views/tree/master/src/main/docs/guide/views

Otherwise looks good!

sgammon commented:
@graemerocher ah yes, those repositories enable the `libs/` folder I was using to inject JARs, specifically, Soy. now that that's cleaned up, I can remove that.

i'll also write up some docs today about how to use it. thanks again, we're really excited to be using micronaut! :)

sgammon commented:
hey @graemerocher -

so, i've cleaned things up and added some docs. here's a screenshot. they include usage, config reference, and examples, including how to use XSS and CSS renaming features.

<img width="905" alt="Screen Shot 2019-09-04 at 5 17 05 PM" src="https://user-images.githubusercontent.com/171897/64302405-13ded680-cf38-11e9-80f6-e399a4005bdf.png">
